File: leetcodepractice/11-container-with-most-water/container-with-most-water.js
/**
 * @param {number[]} height
 * @return {number}
 */

 // [1,8,6,2,5,4,8,3,7]

var maxArea = function(height) {
    // calculate volume
    // l = 0
    // r = height.length - 1
    // compare height between height[l] and height[r] take shorter and multiply that by l - r = area
    let l = 0; // 1
    let r = height.length - 1; //6
    let area = 0; //8

    while (l < r){
        let containerHeight = Math.min(height[l], height[r]); //h[l] = 8, h[r] = 8; = 8
        let base = r - l; //5
        let testArea = containerHeight * base; //40
        if (testArea > area) area = testArea; //49

        if (height[l] < height[r]) { //h[l] = 8, h[r] = 3, false
            l++;
        } else {
            r--;
        }
    }

    return area;
};

File: 11-container-with-most-water/container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        max_water = 0
        left = 0
        right = len(height) - 1

        while left < right:
            length = min(height[left], height[right])
            width = right - left
            area = length * width

            if height[left] <= height[right]:
                left += 1
            else:
                right -= 1

            if area > max_water:
                max_water = area
        
        return max_water